src/openms/source/SYSTEM/NetworkGetRequest.cpp

```diff
@@ -0,0 +1,119 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Timo Sachsenberg $
+// $Authors: Timo Sachsenberg $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/SYSTEM/NetworkGetRequest.h>
+#include <OpenMS/CONCEPT/LogStream.h>
+#include <OpenMS/DATASTRUCTURES/ListUtils.h>
+
+#include <QtGui/QTextDocument>
+#include <QtNetwork/QNetworkRequest>
+
+#include <iostream>
+
+using namespace std;
+
+namespace OpenMS
+{
+
+  NetworkGetRequest::NetworkGetRequest(QObject* parent) :
+    QObject(parent), reply_(0)
+  {
+    manager_ = new QNetworkAccessManager(this);
+  }
+
+  NetworkGetRequest::~NetworkGetRequest()
+  {
+  }
+
+  void NetworkGetRequest::setUrl(QUrl url)
+  {
+    url_ = url;
+  }
+
+  void NetworkGetRequest::run()
+  {
+    if (reply_ == 0)
+    {
+      error_ = QNetworkReply::NoError;
+      error_string_ = "";
+      QNetworkRequest request;
+      request.setUrl(url_);
+      request.setHeader(QNetworkRequest::ContentTypeHeader, "text/plain");
+      connect(manager_, SIGNAL(finished(QNetworkReply*)), this, SLOT(replyFinished(QNetworkReply*)));
+      reply_ = manager_->get(request);
+    }
+  }
+
+  void NetworkGetRequest::replyFinished(QNetworkReply* reply)
+  {
+    if (reply_ != 0)
+    {
+      error_ = reply->error();
+      error_string_ = error_ != QNetworkReply::NoError ? reply->errorString() : "";
+      response_bytes_ = reply->readAll(); // in case of error this will just read the error html from the server
+      reply->close();
+      reply->deleteLater();;
+    }
+    emit done();
+  }
+
+  void NetworkGetRequest::timeOut()
+  {
+    if (reply_ != 0)
+    {
+      error_ = QNetworkReply::TimeoutError;
+      error_string_ = "TimeoutError: the connection to the remote server timed out";
+      reply_->abort();
+      reply_->close();
+      reply_->deleteLater();
+    }
+    emit done();
+  }
+
+  QString NetworkGetRequest::getResponse() const
+  {
+    return QString(response_bytes_);
+  }
+
+  bool NetworkGetRequest::hasError() const
+  {
+    return error_ != QNetworkReply::NoError;
+  }
+
+  QString NetworkGetRequest::getErrorString() const
```

Comment:
This is a good point and we should anyways only count each instance and not
each thread.  Si there should be one call home no matter how many threads

sent from my smartphone -- you may keep any typos you find.
On Apr 29, 2016 12:53 AM, "Timo Sachsenberg" <notifications@github.com>
wrote:

> In src/openms/source/SYSTEM/NetworkGetRequest.cpp
> <https://github.com/OpenMS/OpenMS/pull/1875#discussion_r61544108>:
>
> > +      reply_->deleteLater();
> > +    }
> > +    emit done();
> > +  }
> > +
> > +  QString NetworkGetRequest::getResponse() const
> > +  {
> > +    return QString(response_bytes_);
> > +  }
> > +
> > +  bool NetworkGetRequest::hasError() const
> > +  {
> > +    return error_ != QNetworkReply::NoError;
> > +  }
> > +
> > +  QString NetworkGetRequest::getErrorString() const
>
> so you mean return response_bytes_? I don't know how this is handled but I
> thought this might potentially be not thread safe (e.g. if another request
> is made)
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenMS/OpenMS/pull/1875/files/1fa112200be8ff9d7b56fa290dc64aaf48d9c34d#r61544108>
>